### PR TITLE
Use `targetSchema` of JSON Hyper Schema to communicate sticky action

### PR DESCRIPTION
**Pull request by danielbachhuber**
## Description

Updates the `post-sticky` check to use presence of `wp:action-sticky` in `_links` to determine whether the sticky UI should display. This system is called `targetSchema`. It allows us to compute whether a user can perform an action server-side, and then communicate it to the client.

See https://github.com/WordPress/gutenberg/issues/6361#issuecomment-385769046

## How has this been tested?

For editors and above, the "Stick to the Front Page" UI should display:

image

For authors and below, it should not:

image

[User Switching](https://wordpress.org/plugins/user-switching/) is a helpful plugin for quickly testing this.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

File: editor/components/post-sticky/test/index.js [hunk 1]
```diff
@@ -9,54 +9,51 @@ import { shallow } from 'enzyme';
...
-		},
-	};
-
 	it( 'should not render anything if the post type is not "post"', () => {
 		const wrapper = shallow(
-			<PostStickyCheck postType="page" user={ user }>
+			<PostStickyCheck postType="page" post={ {
+				_links: {
+					self: {
+						href: 'https://w.org/wp-json/wp/v2/posts/5',
```
Comment (kadamwhite) [inline editor/components/post-sticky/test/index.js]: Should this be `/page/5` if `postType="page"`? (Doesn't change test result, admittedly)

File: editor/components/post-sticky/test/index.js [hunk 2]
```diff
@@ -9,54 +9,51 @@ import { shallow } from 'enzyme';
...
 		const wrapper = shallow(
-			<PostStickyCheck postType="post" user={ user }>
+			<PostStickyCheck postType="post" post={ {
+				_links: {
+					self: {
+						href: 'https://w.org/wp-json/wp/v2/posts/5',
+					},
+					'wp:action-sticky': {
+						href: 'https://w.org/wp-json/wp/v2/posts/5',
+					},
```
Comment (kadamwhite) [inline editor/components/post-sticky/test/index.js]: Once again the test works as-is because of what we're testing for, but all of the keys on the `_links` object hold arrays of objects, instead of plain objects (e.g. `wp:action-sticky': [   { /* ... */ }   ]`).

File: lib/rest-api.php [hunk 3]
```diff
@@ -265,6 +265,44 @@ function gutenberg_add_block_format_to_post_content( $response, $post, $request
...
+					'targetSchema' => array(
+						'type'       => 'object',
+						'properties' => array(
+							'sticky' => array(
+								'type'        => 'boolean',
+								'description' => __( 'Whether or not the object should be treated as sticky.', 'gutenberg' ),
+							),
+						),
+					),
+				),
```
Comment (kadamwhite) [inline lib/rest-api.php]: We have this issue all over the API, but this is a lot of data to be piling into the response just to indicate whether the UI can take some action. Do we really want to be repeating all of this:
image
for each post every time we're in a collection where we might _possibly_ want to sticky something? Our `_links` object is getting quite heavy.

I don't see a better option so I'm :+1: on the implementation, but I think we'll need to come back to that at some point.

**Review by kadamwhite — APPROVED** [review]

File: lib/rest-api.php (hunk 3 above)
Comment (TimothyBJacobs) [reply lib/rest-api.php]: If we ever land generating a URL to access a schema, the `targetSchema` section could be replaced with a `$ref` for the property.

Comment (TimothyBJacobs) [reply lib/rest-api.php]: We could also remove the `description` property but it'd make the link less helpful for clients.

File: editor/components/post-sticky/check.js [hunk 4]
```diff
@@ -27,14 +23,8 @@ export function PostStickyCheck( { postType, children, user } ) {
 export default compose( [
 	withSelect( ( select ) => {
 		return {
+			post: select( 'core/editor' ).getCurrentPost(),
```
Comment (gziolo) [inline editor/components/post-sticky/check.js]: It would be better to return only:
```js
hasStickyAction: get( post, [ '_links', 'wp:action-sticky' ], false )`
```
from `withSelect` to avoid unnecessary rerenders when the current post changes after unrelated modifications.

File: editor/components/post-sticky/test/index.js (hunk 1 above)
Comment (danielbachhuber) [reply editor/components/post-sticky/test/index.js]: 53ce334e

File: editor/components/post-sticky/test/index.js (hunk 2 above)
Comment (danielbachhuber) [reply editor/components/post-sticky/test/index.js]: 67f0f9bc87b5bd9c28f1269fb3fe0098490f995b

File: editor/components/post-sticky/check.js (hunk 4 above)
Comment (danielbachhuber) [reply editor/components/post-sticky/check.js]: Good catch, fixed in 3e591c50d08c3128745c44cb7c23d8b8ac991c5f

**Review by felixarntz — DISMISSED** [review]

File: lib/rest-api.php [hunk 5]
```diff
@@ -265,6 +265,44 @@ function gutenberg_add_block_format_to_post_content( $response, $post, $request
...
+ * @return WP_REST_Response Response containing the new links.
+ */
+function gutenberg_add_target_schema_to_links( $response, $post, $request ) {
+	$new_links  = array();
+	$orig_links = $response->get_links();
+	$post_type  = get_post_type_object( $post->post_type );
+	// Only Posts can be sticky.
+	if ( 'post' === $post->post_type ) {
+		if ( current_user_can( $post_type->cap->edit_others_posts )
+			&& current_user_can( $post_type->cap->publish_posts ) ) {
```
[the remark this answers is not in the document]
Comment (danielbachhuber) [reply lib/rest-api.php]: > Why is the capability for publishing posts required to make a post sticky?

I'm using the check within `edit_post()` as my point of reference: https://github.com/WordPress/WordPress/blob/3266b10d04ad43b9d983e0b55d4d6247be33e18a/wp-admin/includes/post.php#L401

> More importantly though, this should use the meta capabilities for the specified post and rely on `map_meta_cap()` instead of handling the logic manually.

I'm not sure I follow. Can you explain why?

[the remark this answers is not in the document]
Comment (danielbachhuber) [reply lib/rest-api.php]: > Therefore I think the check in `edit_post()` is not accurate either. For this very case, there's no appropriate capability yet, so I'd recommend to introduce `stick_post` and `unstick_post` meta capabilities and add a clause for them in `map_meta_cap()`. We can then resolve it to the primitive capabilities we want to in there (if those are decided to be what currently happens in `edit_post()`, that's alright I guess). The benefit is that we then don't have to memorize all over what to check for when making a post sticky or unsticky, we can simply use intuitive `stick_post` / `unstick_post`, both in the current backend, the REST API, anywhere.

Does what you describe need to be solved for in this pull request?

**Review by pento — COMMENTED** [review]
I've been letting this tumble around for a bit, and I have a few concerns.

First off, I like the idea of providing information in the form of "this is an action the current user can perform, and here are the properties that action relates to". This is useful information to provide.

As has been mentioned, this solution seems quite verbose. I guess a major part of that is the `description` being on the post object, instead of in the schema, where all the other descriptions are. It's also duplicating the description of the `sticky` property, rather than documenting what `action-sticky` is for.

On that note, the naming is quite obtuse, too, it took me way too long to understand how the data is supposed to be used. There's nothing to indicate that "the presence of `wp:action-sticky` mean you should render the Sticky Post UI".

I think this would be better in a sibling property to `_links`, `_actions`, or something like that. `_links` is entirely used as a list of shortcuts to related information about the current object, adding in "here is an action the current user can perform" is confusing.

**Comment by kadamwhite** [comment]
> I think this would be better in a sibling property to _links, _actions, or something like that. _links is entirely used as a list of shortcuts to related information about the current object, adding in "here is an action the current user can perform" is confusing.

This would have the added benefit that such information could be omitted using `_fields` when the client does not need to know about actions

**Comment by danielbachhuber** [comment]
> I think this would be better in a sibling property to `_links`, `_actions`, or something like that. `_links` is entirely used as a list of shortcuts to related information about the current object, adding in "here is an action the current user can perform" is confusing.

Can you point me to the existing spec where this is documented?

**Comment by pento** [comment]
> Can you point me to the existing spec where this is documented?

It isn't documented as far as I'm aware, this is just how `_links` seems to be used in the core REST API. It's suggested by the name, too: a link is something you click (or in HTTP verbs, something you `GET`), it's not a thing that you `POST`.

As far as adhering to the JSON API spec goes, I'm fine with using something that exists within the spec, if that fits our requirements. If not, I'm 💯 in favour of doing something that works, rather than something that's shoe-horned into a vaguely related structure.

**Comment by danielbachhuber** [comment]
> It isn't documented as far as I'm aware, this is just how `_links` seems to be used in the core REST API.

`_links` is an implementation of [HAL](http://stateless.co/hal_specification.html). `targetSchema` is a part of the current [JSON Hyper-Schema spec](http://json-schema.org/latest/json-schema-hypermedia.html#rfc.section.6.5.4).

> If not, I'm 💯 in favour of doing something that works, rather than something that's shoe-horned into a vaguely related structure.

I'm not particularly interested in spending a ton of time bikeshedding an alternative spec. Can you describe the alternative implementation you'd accept?

cc @TimothyBJacobs for any thoughts he has.

**Comment by pento** [comment]
Sure, here's a rough idea of something I think would be extensible, would clearly show the difference between `_links` and `_actions`.

In the schema, `_actions` would be defined as an array of all the possible actions, with a description of what they're for, and what they're related to, like so:

```json
{
	"routes": {
		"/wp/v2/posts/(?P<id>[\\d]+)": {
			"endpoints": [
				{
					"methods": [ "POST", "PUT", "PATCH" ],
					"args": {
						// ...
					},
					"_actions": {
						"wp:sticky": {
							"title": "Sticky Post",
							"description": "Whether the current user can sticky the post or not",
							"targetSchema": {
								// Something in here about being related to the "sticky" property.
							}
						}
					}
				},
			],
		},
	},
}
```

Then, in the post object, the `_actions` property would define whether each action is available or not:

```json
{
	"id": 1,
	// ...
	"_actions": {
		"wp:sticky": true,
	},
}
```

Alternatively, `_actions` could be an array of available actions, the difference seems to mostly be a choice of style:

```json
{
	"id": 1,
	// ...
	"_actions": [
		"wp:sticky",
	],
}
```

**Comment by danielbachhuber** [comment]
Again, I'm strongly opposed to inventing a new spec when the existing spec is sufficient. In inventing a new spec, we'll need to go through multiple iterations of relearning past mistake/decision cycles. This doesn't seem like a judicious use of our time.

To address response size concerns, what if we only included these action links if `context=edit`?

**Comment by pento** [comment]
> To address response size concerns, what if we only included these action links if `context=edit`?

Yah, that'd be fine. I don't think there's a need for them in other contexts, we can always add it to them if there is. That's not really the issue, though.

The specific concern I have is that, when reading these new links, there's nothing to suggest what they mean, or how you use them. The same problem exists in @TimothyBJacobs' example: there's nothing to suggest what `wp:action-sticky` means, someone might guess it's related to the `https://api.w.org/action-sticky` thing in the schema, but I wouldn't rely on that. Even if they do, there's still nothing to tell them what it actually means.

The problem we need to solve is that there are pieces of UI that should only be displayed when a certain set of conditions are met: we want to be calculating those conditions on the server, rather than leaving it up to the client. `targetSchema` is nice, but it's a super complex way making available what are essentially a list of flags.

So, here's what I would like to see:

- The post object returning a list of these flags. 
- The schema containing a description of the list, and of each flag.

`_actions` happens to be the name I gave that list in my earlier example, but I have no attachment to that: if there's a spec that more readily fits this, then use that. If `targetSchema` can be adjusted to do those things, then use it.

If there isn't a spec, then we can just make something that works, and iterate it on it later if needed.

**Comment by danielbachhuber** [comment]
@pento I'm not sure I follow your specific recommendation. What do you think about cce57b8eb? Or, if there's other changes you'd like to make, can you communicate them in a pull request against this branch?

**Comment by TimothyBJacobs** [comment]
I disagree. If you understand what `targetSchema` means, then it is immediately obvious what is going on. Gutenberg is filled with dozens of these technical hurdles that aren't immediately obvious to WordPress developers until you understand what a HoC is, or Slot & Fill, etc... The same thing is true with a number of patterns in WordPress core as well. 

I also disagree that it is super complex. Its a fairly pattern of representing what the target resource looks like and is certainly no more complex than many other G7g patterns that are necessary to fulfill a complex requirement.

**Comment by danielbachhuber** [comment]
To clarify my own position: I'm open to being convinced there is an implementation superior to that proposed on this pull request. However, at this point, I'm unclear as to what the alternative implementation is, and how it's functionally better.

From my perspective, the `targetSchema` implementation on this pull request is correct because:

1. It conveys the information needed by the client.
2. It follows an existing, documented spec that much of the REST API already follows.

**Comment by TimothyBJacobs** [comment]
In case it wasn’t clear, I of course have no issue with some other mechanism being used to indicate what actions are available, but I very much do not want to see WordPress adopt another standard in a weird way that isn’t really applicable and break the purpose of it being standard without a very good reason. And I don’t think that `targetSchema` is complicated enough to warrant that breakage.

**Comment by pento** [comment]
Thank you for the discussion, @danielbachhuber and @TimothyBJacobs. cce57b8 is a step in a good direction. I've opened #6613 with some small tweaks, but I think we can go with this.

**Comment by danielbachhuber** [comment]
Thanks @pento.

**Comment by rickalee** [comment]
What is the best method to remove 'Sticky' option/toggle from Posts globally now?
